src/security/collection.cpp

```diff
@@ -0,0 +1,396 @@
+      // if (it.is_delete()) { }
+      // don't reduce the count, we need to allocate space for the provisional
+      // writes (before the deletion)
+      // we could improve this by saving the pointer of the first non-deleted
+      // header per key
+      keys_bag[key]++;
+    }
+
+    // we now know the number of keys; allocate map entries
+    dnsec::ddwaf_obj *entries =
+        memres_.allocate_objects<dnsec::ddwaf_obj>(keys_bag.size());
+    slot.make_map(entries, keys_bag.size());
+    dnsec::ddwaf_obj *next_free_entry = entries;
+
+    // fill the map entries
+    // map that saves the ddwaf_object for keys that occurr more than once
+    std::unordered_map<std::string_view, dnsec::ddwaf_arr_obj *>
+        indexed_entries;
+    for (it.reset(); !it.ended(); ++it) {
+      auto [key, value] = *it;
+      std::size_t const num_occurr = keys_bag[key];
+
+      // common scenario: only 1 occurrence of the key
+      if (num_occurr == 1) {
+        dnsec::ddwaf_obj &entry = *next_free_entry++;
+        entry.set_key(key);
+        entry.make_string(value);
+        continue;
+      }
+
+      auto ie = indexed_entries.find(key);
+      if (ie == indexed_entries.end()) {
+        // first occurrence of this key
+        dnsec::ddwaf_obj &entry = *next_free_entry++;
+
+        entry.set_key(key);
+        auto &arr_val = entry.make_array(num_occurr, memres_);
+        indexed_entries.insert({key, &arr_val});
+
+        if (!it.is_delete()) {
+          arr_val.at_unchecked<dnsec::ddwaf_obj>(0).make_string(value);
+          entry.nbEntries = 1;
+        }
+      } else {
+        // subsequent occurrence of this key
+        auto &arr_val = *ie->second;
+        if (!it.is_delete()) {
+          arr_val.template at_unchecked<dnsec::ddwaf_obj>(arr_val.nbEntries++)
+              .make_string(value);
+        } else {
+          arr_val.nbEntries = 0;
+        }
+      }
+    }
+  }
+
+  static void set_request_uri_raw(const ngx_http_request_t &request,
+                                  dnsec::ddwaf_obj &slot) {
+    set_map_entry_str(slot, kUriRaw, request.unparsed_uri);
+  }
+
+  static void set_request_method(const ngx_http_request_t &request,
+                                 dnsec::ddwaf_obj &slot) {
+    set_map_entry_str(slot, kMethod, request.method_name);
+  }
+
+  // adapt to the same iteratror format as query_string_iter
+  template <bool IsRequest>
+  struct HeaderKeyValueIter {
+    HeaderKeyValueIter(const ngx_list_t &list, std::string_view exclude,
+                       dnsec::DdwafMemres &memres)
+        : list_{list},
+          memres_{memres},
+          exclude_{exclude},
+          it_{list},
+          end_{dnsec::NginxListIter<ngx_table_elt_t>::end(list)} {}
+
+    void reset() { it_ = decltype(it_){list_}; }
+
+    bool ended() const noexcept { return !(it_ != end_); }
+
+    HeaderKeyValueIter &operator++() {
+      while (true) {
+        ++it_;
+        if (!(it_ != end_)) {
```

Comment:
```suggestion
        if (it_ == end_) {
```
The operator exists, better hide the double negative.